Implement readthedocs

Snell1224 opened the pull request:
This PR includes all LDMS-related documentation and Read the Docs (RTD) configuration files, such as .rst files, conf.py, Makefile, and others. It also adds a GitHub Actions workflow, "Build and Trigger Read the Docs", which automates the generation and updating of documentation whenever changes are made to specific files in the docs/ directory.

A couple things to note:
1. RTD Webhook Secret: Someone with admin access to the LDMS repository will need to add a RTD_WEBHOOK_SECRET in the repository's settings. This secret allows GitHub Actions to trigger a build on the Read the Docs site (e.g., [LDMS Documentation](https://ovis-hpc.readthedocs.io/projects/ldms/en/latest/index.html)).
   - NOTE: I’m happy to set this up if I can be temporarily granted admin privileges.
2. .man to .rst Conversion Fixes: Modifications were made to .man pages to resolve errors encountered during the man2rst conversion process.
3. Makefile Updates: The Makefile.am has been updated to generate .man pages from the .rst files (located under docs/man/) within the LDMS build directory.
4. Next Steps: If there are no issues, I’ll apply the same structure and formatting to PRs for the remaining OVIS subprojects.

GITHUB ACTIONS WORKFLOW:
1. Trigger Conditions
  - Runs on **`push`** or **`pull_request`** events when changes occur in:
  - The `docs/` directory or `rst` files.

2. Job1: `generate-rst`
  - Checkout the repository and set up Python 3.8.
  - Install dependencies like `pandoc` for file conversion.
  - Convert `.man` files to `.rst` format, if ``CONVERT_MAN_PAGES`` is set to `true`.
                - Note: This is a temporary step during the transition from .man to .rst documentation. If developers continue to add .man files, they can easily be converted by enabling CONVERT_MAN_PAGES.  
  - Update `index.rst` files dynamically to reflect added/removed or modified `.rst` files.
  - Commit and push changes back to the repository if any documentation files were modified.

3. Job2: `build-docs`
- Depends on the completion of `generate-rst`.
- Prepares the repository for Read the Docs to build the updated documentation.

morrone commented on docs/asf/pyanalysis.rst:
```diff
@@ -0,0 +1,298 @@
+
+grafanaFormatter:
+
+.. code:: RST
```
Shouldn't this be "python" instead of RST? The same for number of the following code blocks.

Snell1224 replied:
Yes, you are correct. It has been fixed, thanks for the catch!

morrone commented on docs/container-quickstart.rst:
```diff
@@ -0,0 +1,931 @@
+     }
+   }
+
+Maestro Container
```
Maybe this should be documented in the maestro project instead of here in the ldms project?

morrone commented on docs/container-quickstart.rst:
```diff
@@ -0,0 +1,931 @@
+Then, you should be able to access the grafana web server via
+``http://127.0.0.1:3000/`` on your laptop.
+
+Building Containers
```
Maybe this documentation belongs in the ovis-hpc/ldms-containers project rather than here in the ovis-hpc/ldms project?

Snell1224 replied:
Oh, I was not aware of the ldms-containers subproject. This .rst file is the same as the README.md for this repo so I will go ahead and remove it.

morrone commented on docs/files/ldmscon2023_directory.zip:
Please do not include zip files in the ldms git repo.

morrone reviewed (changes requested):
This looks like a good start!

If we are going to have all of the rst files checked into the tree, then I think we need to remove the man files. We don't want to have to deal with keeping two different formats of the same documentation updated. On or the other should be auto-generated. I'd vote for  switching to rst.

I'm not at all a fan of sprinkling index.rst files throughout the tree. I don't think the documentation and man pages would necessarily need to mimic the code tree structure, especially for man pages. I think we should have co-located high level files that contain the document structure, and has pointers to the various documents it links to. 

If we move to .rst files and delete the .man files, then we need to make rst2man a requirement in configure.ac. We probably need to iterate a bit on tree structure.

Snell1224 commented:
@morrone Thanks for reviewing this! I've updated the branch with the suggested changes but haven’t modified the tree structure under docs/man/ yet. It seems that most of us are on board with transitioning to .rst, and I agree that we should discuss the best way to organize these documents (maybe during the next LDMS UG meeting) before removing any existing .man files. @tom95858, do you have any additional thoughts or suggestions?

baallan commented:
I have no strong feelings about summary and umbrella documentation locations, but the .rst file(s) specific to a plugin should live in the directory with the plugin. This allows the contributor of a plugin to work in a clearly defined sandbox.

morrone commented:
> I have no strong feelings about summary and umbrella documentation locations, but the .rst file(s) specific to a plugin should live in the directory with the plugin. This allows the contributor of a plugin to work in a clearly defined sandbox.

I agree. I just want to get rid of the "index.rst" files that are strewn throughout, not the content rst files (which would be the man page source as well.

baallan commented:
We might also take this opportunity to put some consistency in place wrt man page naming and to drop/change the 'plugin' prefix on our plugin man pages.  As complex tooling manuals go, the pattern for git man page naming is one of the least unpleasant I've run into in quite a while.
Ideally 'man ldms' gets you an index page.

baallan commented:
I would like all the following to return something sensible (note the lack of a pedantic plugin type namespacing):
```
man ldms   (index)
man ldms-plugins (plugin index: covers auth, zap, sampler, store, and stream plugin lists; possibly also list man pages for C multi-instance plugin apis)
man ldms-meminfo (the meminfo plugin man page)
man ldms-munge (the munge auth plugin man page)
```
and I would *eventually* like
```
man ldms-genders
```
to return:
```
 It's gone. Muahahahaha!
```

Snell1224 commented:
@morrone @baallan Thanks for the suggestions and feedback!

I've removed the extra index.rst files that were scattered throughout the docs/man/ directories and reorganized the structure in a way that made the most sense to me. If you have a preferred tree structure in mind, let me know—I'm happy to change it!

I also added `ldms-idx.man`, which lists all current man pages (generated in docs/Makefile.am). If we like it, great! I can also add an `ldms-plugins-idx.man` to list only the sampler/storage plugins too. If not then I have no issues with removing it.

tom95858 commented:
@baallan, @morrone, @Snell1224 are we happy with this? I would like to merge it if you all feel that it is a good start.

tom95858 commented:
@Snell1224 Prior to merge, it looks like there is a trivial conflict in docs/Makefile.am that needs to get resolved.

morrone commented:
> @baallan, @morrone, @Snell1224 are we happy with this? I would like to merge it if you all feel that it is a good start.

I think it is making good progess, but needs at least one thing addressed before landing. We either need to remove the static man pages in favor of the new automatically generated man pages, or we need to _not_ install (at "make install" time) the automatically generated man pages. We should probably not land it while we have both static and dynamic man pages competing to install different versions of the same target file.

And I made a couple of smaller comments inline that are easy to address.

I would like us to make a call on making rst2man becoming an ldms build requirement rather than being optional. I vote for making it a hard requirement.

Snell1224 commented on docs/Makefile.am:
```diff
@@ -1,17 +1,51 @@
-EXTRA_DIST += ldms_stream.rst
-man7_MANS += ldms_stream.man
+EXTRA_DIST += $(RST_FILES)
+dist_man1_MANS = $(call categorize_man_pages,1)
```
I agree. I wasn't sure if we wanted to keep the conditional statement so I just changed the AC_MSG from warning to error in configure.ac and left everything else as is.

Snell1224 commented:
I am happy with where we are at now and agree that it'd be best to make rst2man a requirement (which I included in the most recent commit). As for whether or not we keep the static man pages, I vote for removing them and from here on out stick to RST documentation.

morrone commented:
I did a test build and "make install" with this PR. Still moving in a good direction! But we have some things that still need work:

- [ ] Some of the new man page file names are being generated with duplicate section numbers. For example, "procnetdev2.7.7".
- [ ] The new man pages are being generated with duplicate "NAME" sections, and the metadata for the man pages are wrong. For example:
```
PROCNETDEV2(07 Jan 2022)                                                                    PROCNETDEV2(07 Jan 2022)

NAME
       procnetdev2 - -

NAME
       procnetdev2 - man page for the LDMS procnetdev2 plugin
```
- [ ] We need .gitignore files that list each generated .man file (so they don't show up in "git status" after doing a "make")
- [ ] The rst man page sources need to be distrbuted throughout the code rather than all in one place (e.g. docs/man/plugins/sampler/procnetdev2/procnetdev2.7.rst should be moved to ldms/src/sampler/procnetdev2/procnetdev2.7.rst)
- [ ] I don't think that we want to have duplicate man pages installed, so we should remove the old man pages (e.g. remove Plugin_procnetdev2.man in favor of procnetdev2.7.rst)

morrone commented:
On the topic of the duplicated "NAME" sections and faulty man page metadata, I think this is the example that we should follow to fix our issues:

https://docutils.sourceforge.io/sandbox/manpage-writer/rst2man.txt

Notice that is doesn't contain an explicit "NAME" section. rst2man generates the NAME section by combining the main title and the test between the lines of minux signs (-). For metadata, we'll want at least the "Manual section" and "Manual group". We probably don't need "Author" for instance.

That example also shows a way to do the "SEE ALSO" section that won't need a bunch of extra processing. It may not be exactly like we are used to, but I think it is probably easiest to just go with something like that example.

Snell1224 commented:
I am working on getting the generated man pages fixed (thanks for catching this!), adding .gitignore for them and will be removing the original man pages. 

As for the .rst file location, I thought we had agreed that keeping them all under docs/ would be the best approach? Centralizing them makes maintenance and organization much easier, especially with a single Makefile.am to generate the man pages. 

I do see the benefit of having .rst files alongside the source code for developer convenience, but I think the trade-off in maintainability outweighs that. @tom95858 @baallan @narategithub What does everyone else think? If most prefer distributing the man page .rst files across the code, I’ll go ahead and make the changes.

baallan commented:
I think plugin doc source and code source need to be in the same file location, as updating one usually implies updating the other is needed and it should be trivial to find the paired files.
If there is a concern about defining rules for rst processing in makefile.am, one can use find . -name '*.rst' in a higher level makefile to gather all inputs for rst2man.

morrone commented:
> As for the .rst file location, I thought we had agreed that keeping them all under docs/ would be the best approach? Centralizing them makes maintenance and organization much easier, especially with a single Makefile.am to generate the man pages.

I don't think we had.  I had agreed with @baallan that the .rst files that go with plugins, and are used to generate the man pages, should be in the directories with the plugins. I just didn't want the index.rst files all over the place. But I think we would like to have the rst files that generate man pages co-located with the code that they document.

I think that if we follow the example rst man page, we can get rid of the complicated regexes that were needed to rewrite the rst on the fly. Then it will be simple to have the rst2man rules in multiple Makefile.am.

Snell1224 commented:
Oh right! That was mistake on my part, I didn’t remember the initial suggestion on keeping the .rst files in the same directory as the source code. I’ll go ahead and make these changes :)

narategithub commented:
I think rst docs should be in top-dir `docs/`. If they are spreading out to in `ldms/src/`, I think it will complicate read-the-docs build. For example, [flux-core](https://github.com/flux-framework/flux-core) put all RST files in `docs/`.

morrone commented:
I just realized something else we need to fix. We need to come up with some sane filename prefix for all of the plugin man pages. I don't think it is reasonable for ldms to install very generic man pages named things like "edac", "papi, "loadavg", "lustre_client", etc. We need to make it clear that these are part of ldms. Previously we had "Plugin_", but that wasn't terribly good either.

Since the plugins are for ldmsd, I'll suggest the prefix "ldmsd_". Other man pages might be prefixed with "ldms_" as approriate. I can't usually tell which things are ldms and which are ldmsd.

tom95858 commented:
I agree with @morrone. I don't have strong feelings about the convention,
but I absolutely agree with Chris that we have namespace pollution issues
that we should resolve now.


On Thu, Mar 6, 2025 at 12:33 PM Christopher J. Morrone <
***@***.***> wrote:

> I just realized something else we need to fix. We need to come up with
> some sane filename prefix for all of the plugin man pages. I don't think it
> is reasonable for ldms to install very generic man pages named things like
> "edac", "papi, "loadavg", "lustre_client", etc. We need to make it clear
> that these are part of ldms. Previously we had "Plugin_", but that wasn't
> terribly good either.
>
> Since the plugins are for ldmsd, I'll suggest the prefix "ldmsd_". Other
> man pages might be prefixed with "ldms_" as approriate. I can't usually
> tell which things are ldms and which are ldmsd.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ovis-hpc/ldms/pull/1599#issuecomment-2705090297>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVTPXGNK4UAB4EFUPHALYL2TDEK5AVCNFSM6AAAAABWPMWVPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBVGA4TAMRZG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> [image: morrone]*morrone* left a comment (ovis-hpc/ldms#1599)
> <https://github.com/ovis-hpc/ldms/pull/1599#issuecomment-2705090297>
>
> I just realized something else we need to fix. We need to come up with
> some sane filename prefix for all of the plugin man pages. I don't think it
> is reasonable for ldms to install very generic man pages named things like
> "edac", "papi, "loadavg", "lustre_client", etc. We need to make it clear
> that these are part of ldms. Previously we had "Plugin_", but that wasn't
> terribly good either.
>
> Since the plugins are for ldmsd, I'll suggest the prefix "ldmsd_". Other
> man pages might be prefixed with "ldms_" as approriate. I can't usually
> tell which things are ldms and which are ldmsd.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ovis-hpc/ldms/pull/1599#issuecomment-2705090297>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABVTPXGNK4UAB4EFUPHALYL2TDEK5AVCNFSM6AAAAABWPMWVPSVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDOMBVGA4TAMRZG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>


-- 
Thomas Tucker, President, Open Grid Computing, Inc.

tom95858 commented:
Here's a start:

LDMSD Plugins:
* ldmsd_store_xxx
* ldmsd_sampler_***
* ldmsd_auth_***
* ldmsd_this
* ldmsd_that

LDMS Plugins:

* ldms_xprt_rdma
* ldms_xprt_sock
* ldms_xprt ...

LDMS Commands/API

* ldms_...
* 
Please offer alternatives...

tom95858 commented:
From my perspective, the nice thing about keeping the doc with the code is that we where the documentation is. The bad thing(s) about keeping the doc with the code are that:

* We don't know what's not documented
* We can't conveniently make 'global' changes without some find search
* It seems like maintaining indexes, TOC, etc... are easier and more intuitive when all the links you are maintaining are in one place.
* It is immediately obvious if a doc file is not maintaining the file naming conventions

baallan commented:
re man page naming:
- I agree that Plugin_  was never a great idea.
- man allows aliasing of a given page, so we don't need to break the muscle memory of existing admins reading stuff under the current names. 
- all pages, except 1 index page for ldmsd itself about daemon invocation command line and env vars, should start with 'ldms-'. Thus 'man ldms-'{tab} gives the user the index of everything.
- Of the examples of vast man page collections available (perl and git), I *highly* recommend following the git convention, e.g.
```
git
git-commit
git-diff
```
where index pages have whatever hyphenated names and plugins with single functionality look like
 (getting rid of the unpronounceable 'xprt' and poorly RSI-inducing '_')
```
ldms-sampler-meminfo
ldms-auth-ovis
ldms-conn-sock
ldms-conn-rdma
ldms-transform-foobar
```
and plugins with multiple functionality (stream + metric set, various transforms, etc) are simply
```
ldms-plugin-xyzpdq
```
For configuration syntaxes, we should avoid creating enormous numbers (eg 1 per scripting command) of man pages. Keep configurating info together by language:
```
ldms-yaml
ldms-script
ldms-python (alias of ldmsd_controller)
```

Snell1224 commented:
All man pages have been removed the .rst files have been renamed and moved to the corresponding code directory (i.e. store_flatfile.rst --> ldms/src/store/store_flatfile/ldms-store-flatfile.rst). The Makefiles, docs/ and .gitignore have been updated accordingly and I'll be working on debugging the current distcheck errors. In the meantime, please let me know if there are any updates/fixes that need to be made.

morrone commented:
> All man pages have been removed the .rst files have been renamed and moved to the corresponding code directory (i.e. store_flatfile.rst --> ldms/src/store/store_flatfile/ldms-store-flatfile.rst). The Makefiles, docs/ and .gitignore have been updated accordingly and I'll be working on debugging the current distcheck errors. In the meantime, please let me know if there are any updates/fixes that need to be made.

You probably already know this, but just in case, It looks like the rst2man command is missing in a number of the container images. We'll need to have the container images updated.

But the missing rst2man command should have been detected a configure time and thrown an error. I think we lost the configure.ac change that uses AC_MSG_ERROR for a missing rst2man.

morrone commented on ldms/rules.mk:
```diff
@@ -0,0 +1,24 @@
+# 1. Fix duplicate titles when converting to man
+# 	- Remove "NAME" and "=====" underneath
```
Why are we doing this dynamically? Why don't we just fix all the rst sources to use the following format?

https://docutils.sourceforge.io/sandbox/manpage-writer/rst2man.txt

morrone commented on ldms/rules.mk:
```diff
@@ -0,0 +1,24 @@
+# 1. Fix duplicate titles when converting to man
+# 	- Remove "NAME" and "=====" underneath
+# 	- Add :title: to first sentence after :Date:
+# 2. Remove :ref:`<>` links in .rst to avoid rs2man "Unkown interpretd text role "ref"" errors
```
Could we adopt the formatting used in the example:

https://docutils.sourceforge.io/sandbox/manpage-writer/rst2man.txt

and avoid this too?

morrone commented on ldms/rules.mk:
```diff
@@ -0,0 +1,24 @@
+# 	Ex: :ref:`ldms_quickstart(7) <ldms_quickstart>` --> ldms_quickstart(7).
+# 3. Remove generated "/" after every special character.
+# 	Ex: \fB\-a,\fP\fI\-\-default_auth\fP --> \fB-a,\fP\fI--default_auth\fP
+# 4. Remove lines with unnecessary rst2man output.
```
Were they causing problems? I'd love to see us able to get rid of all the sed usage.

Snell1224 replied:
Not sure, I've done some research on this and from what I've found, fst documentation wasn't designed/targeted to be built as man pages.

morrone commented on docs/rst_man/src/sampler/filesingle/ldms-sampler-filesingle.rst:
```diff
@@ -0,0 +1,129 @@
+filesingle
+=================
+
+:Date:   15 Dec 2018
```
I think we need the following in each man page file as well:

:Manual section: 
:Manual group:

Snell1224 commented:
> AC_MSG_ERROR

nichamon commented:
@Snell1224  I built and installed the code based on this PR. The man pages are renamed by replacing `_` with `-`. It makes sense for `ldmsd-failover` and `ldmsd-decomposition` because the man pages describe how the failover and decomposition work. However, `ldmsd-controller` shouldn't be renamed because the man pages describes how `ldmsd_controller` python interface to `ldmsd` work. `ldmsd_controller` is the program name, so the man page should have the same name. The other three mange pages that fall in the same situation are `ldmsd-stream-publish`, `ldmsd-stream-subscribe`, and `ldmsd-yaml-parser`. There may be more but these are what I found so far.

Thank you for all the work! I understand there are many components involved in making this work complete.

Snell1224 commented:
- Updated the man page filenames to make them consistent with the code. All names now follow the format: "ldms", "ldmsd", "ldms-sampler", or "ldms-store", followed by an underscore and the original plugin/code name (e.g., ldms-sampler_procstat2.rst, ldmsd_controller.rst, ldms_auth_none.rst). 
- Added a formatting example at the beginning of the pages: [RST to Man Example](https://docutils.sourceforge.io/sandbox/manpage-writer/rst2man.txt).
- Also, moved the .rst file copy process from Makefile.am to the GitHub workflow, so users/developers won't need to compile and manually commit updated .rst files.

Please let me know if any additional changes need to be made.

morrone commented on docs/rst_man/src/contrib/sampler/daos/ldms-sampler_daos_sampler.rst:
```diff
@@ -5,10 +5,9 @@ daos_sampler
 ===================
 
 :Date:   28 Apr 2022
```
In all of the man pages, we need to move these metadata field lists need to be moved to _after_ the subtitle like they are in https://docutils.sourceforge.io/sandbox/manpage-writer/rst2man.txt. They aren't converting to man pages correctly as they currently are.

Snell1224 replied:
Ah, I was following the ldmsd_decomposition.rst formatting so thank you for pointing this out! Man pages should be correct now.

tom95858 commented:
Hi @Snell1224 , I'm so sorry but it appears that the dcgm change introduced a merge conflict that must be resolved. Could you please fix the merge conflict and I will merge this right away.

Snell1224 commented:
The merge conflict has been fixed and the .rst file of the dcgm man page has been updated. Thank you for letting me know!

tom95858 commented:
@Snell1224 could you please check the pull request. I'm still seeing build failures and cannot merge.

Snell1224 commented:
@tom95858 The build failures are because `rst2man` is not installed on the containers. The configure.ac`  has been updated to throw an error instead of a warning when the `RST2MAN` condition isn't met. I did do a local build/install and distcheck on the current commit and didn't come across any issues or errors.

Snell1224 commented:
> Commit [0fe1e0b](https://github.com/ovis-hpc/ldms/commit/0fe1e0bd7e9b6928133bf603dfe1c334d9e5c7c4) is labelled "merge fixes", and it certainly looks like a merge commit in content. But it isn't showing up as a merge commit to git. I am curious how that could happen? Do you recall what steps you used to make that commit? Were patches generated and applied manually instead of using the "git merge" command?

The commit doesn’t show as a merge because I had a conflict with `Plugin_dcgm_sampler.man`. It was modified in the main branch but deleted in mine. During the merge, git detected the conflict which I resolved by keeping my change (i.e. deleting the file). However, git showed that my branch was `already up to date`, probably because there were no further changes for git to apply so I think it treated the merge as unnecessary.

As a work around, I used `git merge --squash` and manually committed (git commit -m "merge fixes") and pushed the changes. I think this is why it looks like a regular commit instead of a merge.

Thanks for the information about the last commit! I'll keep that in mind

tom95858 commented:
@narategithub, could you please install rst2man on the build containers so that we can accept Sarah's change?

narategithub commented:
@tom95858 now all build containers (`ovishpc/ldms-dev`, `ovishpc/ovis-centos-build`, `ovishpc/ovis-ubuntu-build`, and `ovishpc/ovis-compat-centos7`) have rst2man.

@Snell1224 could you please force-push your branch so that it triggers the build test?

Snell1224 commented:
@narategithub It looks like some of the builds were cancelled due to a `scheduled Ubuntu 20.04 brownout`.

Snell1224 commented:
@tom95858 @narategithub Made some minor updates to the github workflows so the builds will run. Looks like everything is ready to be merged!

tom95858 commented:
@Snell1224 can you please fix the merge conflicts?

tom95858 commented:
> @Snell1224 can you please fix the merge conflicts?

Hi Sarah, can you please fix the merge conflicts?

Snell1224 commented:
@tom95858 I'm not sure what the conflicts are. From my end it says that there are "no conflicts with base branch"

morrone commented:
> > @Snell1224 can you please fix the merge conflicts?
> 
> Hi Sarah, can you please fix the merge conflicts?

I am not sure what merge conflicts you are referring to either. Github says there are no merge conflicts, and looking at the trees, it looks like commit 6e033181d29bd0012f7fc3274fff5e24e566340d is a clean fast-forward merge of main.

@tom95858 , maybe you need to update your copy of "main"? Or you have a newer main that you haven't pushed?

But it would be nice to see this one rebased and squashed to clean up the history.

nichamon commented:
@Snell1224 @morrone @tom95858 This is what I saw on my screen which confuses me a lot. It indicated these two sentences: "This branch cannot be rebased due to conflicts" and "Changes can be cleanly merged."

![image](https://github.com/user-attachments/assets/612fd41a-d8f8-4824-b9ef-dc9260a35b73)

Snell1224 commented:
@morrone @tom95858 @nichamon Got it thank you! I'll go ahead and rebase it.
